### ISSUE TITLE
Exception when saving

### ISSUE THREAD

**Issue by alonaviram** [CarnegieTechnologies/gallery_saver#152]
gallery_saver: ^2.2.0
compileSdkVersion 30
targetSdkVersion 30

```
Fatal Exception: java.lang.NullPointerException
       at carnegietechnologies.gallery_saver.GallerySaver.finishWithSuccess(GallerySaver.kt:77)
       at carnegietechnologies.gallery_saver.GallerySaver.access$finishWithSuccess(GallerySaver.kt:16)
       at carnegietechnologies.gallery_saver.GallerySaver$saveMediaFile$1.invokeSuspend(GallerySaver.kt:72)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:236)
       at android.app.ActivityThread.main(ActivityThread.java:7864)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:620)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1011)

```
can't reproduce on my device but i can see many exceptions like this in crashlytics.. any idea?
thanks!

**Comment by lecicdjuro** [comment]
@alonaviram   I will investigate this next week, or if you have time, please make a pr

**Comment by lecicdjuro** [comment]
Fxied

**Comment by alonaviram** [comment]
the crash is still occurring in 2.3.0.
can you please re-open issue? @lecicdjuro 

`
Fatal Exception: java.lang.NullPointerException
       at carnegietechnologies.gallery_saver.GallerySaver.finishWithSuccess(GallerySaver.kt:79)
       at carnegietechnologies.gallery_saver.GallerySaver.access$finishWithSuccess(GallerySaver.kt:16)
       at carnegietechnologies.gallery_saver.GallerySaver$saveMediaFile$1.invokeSuspend(GallerySaver.kt:74)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
       at android.os.Handler.handleCallback(Handler.java:888)
       at android.os.Handler.dispatchMessage(Handler.java:100)
       at android.os.Looper.loop(Looper.java:213)
       at android.app.ActivityThread.main(ActivityThread.java:8178)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1101)
`

**Comment by cihadturhan** [comment]
Hi, I think my issue is duplicate of this (I don't know why it shows java lines) See also https://github.com/CarnegieTechnologies/gallery_saver/issues/161

**Comment by GoldenCompany** [comment]
Hello,

I'm using the 2.3.2, and i'm having the same issue when i try to save multiple videos at the same time. One video works fine, two videos with enough time between the dowloads is fine.
But two videos (maybe more) in the same period cause the issue.
The first video is always downloaded, the second sometime.

**Comment by carlosfiori** [comment]
Same here, but with pictures.
If I try to save multiple images in sync they work fine.
But if it's in parallel, it crashes the app.

Works:
```dart
 
static Future<List<bool?>> downloadAllImages(List<String> urls) async {
  final distinctUrls = urls.toSet().toList();

  final results = <bool?>[];
  for (var url in distinctUrls) {
    final result = await GallerySaver.saveImage(url);
    results.add(result);
  }
  return results;
 }

```

Crashs the app:
```dart
static Future<List<bool?>> downloadAllImages(List<String> urls) async {
  final distinctUrls = urls.toSet().toList();

  return Future.wait(distinctUrls.map(GallerySaver.saveImage));
}
```
```
E/AndroidRuntime( 7464): java.lang.NullPointerException
E/AndroidRuntime( 7464): 	at carnegietechnologies.gallery_saver.GallerySaver.finishWithSuccess(GallerySaver.kt:79)
E/AndroidRuntime( 7464): 	at carnegietechnologies.gallery_saver.GallerySaver.access$finishWithSuccess(GallerySaver.kt:16)
E/AndroidRuntime( 7464): 	at carnegietechnologies.gallery_saver.GallerySaver$saveMediaFile$1.invokeSuspend(GallerySaver.kt:74)
E/AndroidRuntime( 7464): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
E/AndroidRuntime( 7464): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
E/AndroidRuntime( 7464): 	at android.os.Handler.handleCallback(Handler.java:883)
E/AndroidRuntime( 7464): 	at android.os.Handler.dispatchMessage(Handler.java:100)
E/AndroidRuntime( 7464): 	at android.os.Looper.loop(Looper.java:214)
E/AndroidRuntime( 7464): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
E/AndroidRuntime( 7464): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 7464): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/AndroidRuntime( 7464): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
I/Process ( 7464): Sending signal. PID: 7464 SIG: 9
Lost connection to device.
```

**Comment by rensvis** [comment]
I have exactly the same behavior as @carlosfiori. Would be great if there was a method that took a list of paths/urls.

**Comment by dhruvinnv** [comment]
I was facing the same issue, So tried to solve this by lock mechanism : 


```dart
import 'package:gallery_saver/gallery_saver.dart';
import 'package:synchronized/synchronized.dart';

/// A wrapper class aground image saving process for preventing the deadlocks
/// on the plugin side.

class GallerySaverWrapper {
  GallerySaverWrapper._() : _lock = Lock();
  static final _instance = GallerySaverWrapper._();
  static GallerySaverWrapper get instance => _instance;

  final Lock _lock;

  /// Helps to complete [_saveImageToGallery] in sync.
  Future<void> saveImageToGallery(
    String filePath, {
    String? albumName,
  }) async {
    await _lock.synchronized(
      () => _saveImageToGallery(filePath, albumName: albumName),
    );
  }

  /// Saves the image to the gallery.
  Future<void> _saveImageToGallery(String filePath, {String? albumName}) async {
    await GallerySaver.saveImage(
      filePath,
      albumName: albumName,
    );
  }
}
```
